src/DataStreams/ConnectionCollector.cpp

```diff
@@ -0,0 +1,106 @@
+#include <DataStreams/ConnectionCollector.h>
+
+#include <Core/BackgroundSchedulePool.h>
+#include <Interpreters/Context.h>
+#include <Common/Exception.h>
+#include <common/logger_useful.h>
+
+namespace CurrentMetrics
+{
+extern const Metric AsyncDrainedConnections;
+extern const Metric SyncDrainedConnections;
+}
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int UNKNOWN_PACKET_FROM_SERVER;
+}
+
+std::unique_ptr<ConnectionCollector> ConnectionCollector::connection_collector;
+
+ConnectionCollector::ConnectionCollector(ContextMutablePtr global_context_, size_t max_threads)
+    : WithMutableContext(global_context_), pool(max_threads, max_threads, max_threads * 10)
+{
+}
+
+ConnectionCollector & ConnectionCollector::init(ContextMutablePtr global_context_, size_t max_threads)
+{
+    if (connection_collector)
+    {
+        throw Exception("Connection collector is initialized twice. This is a bug.", ErrorCodes::LOGICAL_ERROR);
+    }
+
+    connection_collector.reset(new ConnectionCollector(global_context_, max_threads));
+    return *connection_collector;
+}
+
+void ConnectionCollector::enqueueConnectionCleanup(std::unique_ptr<IConnections> connections, bool sync) noexcept
```

Comment:
I don't like `sync` flag here. Can we directly call `drainConnections` instead of using it?